Add `Iterator::map_while`

WaffleLapkin opened the pull request:
In `Iterator` trait there is `*_map` version of [`filter`] — [`filter_map`], however, there is no `*_map` version of [`take_while`], that can also be useful.

### Use cases
In my code, I've found that I need to iterate through iterator of `Option`s, stopping on the first `None`. So I've written code like this:
```rust
let arr = [Some(4), Some(10), None, Some(3)];
let mut iter = arr.iter()
    .take_while(|x| x.is_some())
    .map(|x| x.unwrap());

assert_eq!(iter.next(), Some(4));
assert_eq!(iter.next(), Some(10));
assert_eq!(iter.next(), None);
assert_eq!(iter.next(), None);
```
Thit code
1) isn't clean 
2) In theory, can generate bad bytecode (I'm actually **not** sure, but I think that `unwrap` would generate additional branches with `panic!`)

The same code, but with `map_while` (in the original PR message it was named "take_while_map"):
```rust
let arr = [Some(4), Some(10), None, Some(3)];
let mut iter = arr.iter().map_while(std::convert::identity);
```

Also, `map_while` can be useful when converting something (as in [examples]).

[`filter`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter
[`filter_map`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map
[`take_while`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while
[examples]: https://github.com/rust-lang/rust/compare/master...WaffleLapkin:iter_take_while_map?expand=1#diff-7e57917f962fe6ffdfba51e4955ad6acR1042

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @sfackler (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13977&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13977/logs/62), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13977/logs/62)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-20T16:02:59.1984724Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-20T16:03:00.0431378Z ##[command]git config gc.auto 0
2019-11-20T16:03:00.0435518Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-20T16:03:00.0440007Z ##[command]git config --get-all http.proxy
2019-11-20T16:03:00.0442520Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66577/merge:refs/remotes/pull/66577/merge
---
2019-11-20T16:08:40.8537373Z    Compiling serde_json v1.0.40
2019-11-20T16:08:42.6561840Z    Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
2019-11-20T16:08:52.4864382Z     Finished release [optimized] target(s) in 1m 20s
2019-11-20T16:08:52.4959577Z tidy check
2019-11-20T16:08:53.3217507Z tidy error: /checkout/src/libcore/iter/traits/iterator.rs:7: line longer than 100 chars
2019-11-20T16:08:53.3236806Z tidy error: /checkout/src/libcore/iter/traits/iterator.rs: too many lines (3091) (add `// ignore-tidy-filelength` to the file to suppress this error)
2019-11-20T16:08:54.9859518Z some tidy checks failed
2019-11-20T16:08:54.9860731Z Found 441 error codes
2019-11-20T16:08:54.9861215Z Found 0 error codes with no tests
2019-11-20T16:08:54.9861484Z Done!
2019-11-20T16:08:54.9861484Z Done!
2019-11-20T16:08:54.9861670Z 
2019-11-20T16:08:54.9861852Z 
2019-11-20T16:08:54.9862861Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor"
2019-11-20T16:08:54.9863399Z 
2019-11-20T16:08:54.9863599Z 
2019-11-20T16:08:54.9865764Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2019-11-20T16:08:54.9866051Z Build completed unsuccessfully in 0:01:23
2019-11-20T16:08:54.9866051Z Build completed unsuccessfully in 0:01:23
2019-11-20T16:08:54.9916763Z == clock drift check ==
2019-11-20T16:08:54.9926882Z   local time: Wed Nov 20 16:08:54 UTC 2019
2019-11-20T16:08:56.5153482Z   network time: Wed, 20 Nov 2019 16:08:56 GMT
2019-11-20T16:08:56.5153590Z == end clock drift check ==
2019-11-20T16:08:57.8028650Z 
2019-11-20T16:08:57.8125032Z ##[error]Bash exited with code '1'.
2019-11-20T16:08:57.8154642Z ##[section]Starting: Checkout
2019-11-20T16:08:57.8156372Z ==============================================================================
2019-11-20T16:08:57.8156422Z Task         : Get sources
2019-11-20T16:08:57.8156485Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

sfackler commented:
That method name is a lot, but it does seem potentially useful?

@rfcbot fcp merge

rkruppe commented:
For the record, this was previously proposed in https://github.com/rust-lang/rust/pull/65864. It was not merged because I (randomly assigned reviewer) and the two T-libs members who voiced an opinion after my ping were lukewarm on adding it to std instead of itertools. Take from that what you will.

WaffleLapkin commented:
> For the record, this was previously proposed in #65864.

Oh. When I first started thinking of `take_while_map` the #65864 didn't exist yet, so when I finally started implementing this, I didn't search for similar PRs :sweat_smile:

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13979&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13979/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13979/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-20T16:46:43.5713784Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-20T16:46:43.5902895Z ##[command]git config gc.auto 0
2019-11-20T16:46:43.5981240Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-20T16:46:43.6036648Z ##[command]git config --get-all http.proxy
2019-11-20T16:46:43.6185233Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66577/merge:refs/remotes/pull/66577/merge
---
2019-11-20T17:44:31.7630761Z .................................................................................................... 1500/9257
2019-11-20T17:44:37.5221004Z .................................................................................................... 1600/9257
2019-11-20T17:44:45.9818912Z .................................................................................................... 1700/9257
2019-11-20T17:44:54.8031210Z ........i........................................................................................... 1800/9257
2019-11-20T17:45:01.1548601Z .............................................................................................iiiii.. 1900/9257
2019-11-20T17:45:22.9275058Z .................................................................................................... 2100/9257
2019-11-20T17:45:25.3964813Z .................................................................................................... 2200/9257
2019-11-20T17:45:28.0215234Z .................................................................................................... 2300/9257
2019-11-20T17:45:34.2310297Z .................................................................................................... 2400/9257
---
2019-11-20T17:48:23.1748153Z ..............................................................................................i..... 4700/9257
2019-11-20T17:48:29.3221867Z ..........i......................................................................................... 4800/9257
2019-11-20T17:48:38.6455528Z .................................................................................................... 4900/9257
2019-11-20T17:48:43.6645650Z .................................................................................................... 5000/9257
2019-11-20T17:48:53.5826069Z ...................................................................................................i 5100/9257
2019-11-20T17:48:58.2967687Z i.ii...........i.................................................................................... 5200/9257
2019-11-20T17:49:07.8302726Z .................................................................................................... 5400/9257
2019-11-20T17:49:17.8872686Z .................................................................................i.................. 5500/9257
2019-11-20T17:49:25.6117283Z .................................................................................................... 5600/9257
2019-11-20T17:49:31.6284093Z .................................................................................................... 5700/9257
2019-11-20T17:49:31.6284093Z .................................................................................................... 5700/9257
2019-11-20T17:49:41.6692653Z ...................................................................ii...i..ii...........i........... 5800/9257
2019-11-20T17:50:03.3575499Z .................................................................................................... 6000/9257
2019-11-20T17:50:11.8691906Z .................................................................................................... 6100/9257
2019-11-20T17:50:11.8691906Z .................................................................................................... 6100/9257
2019-11-20T17:50:16.5102664Z ........................................................................................i..ii....... 6200/9257
2019-11-20T17:50:40.8725552Z .................................................................................................... 6400/9257
2019-11-20T17:50:48.9960026Z ........................................................i........................................... 6500/9257
2019-11-20T17:50:51.0637128Z .................................................................................................... 6600/9257
2019-11-20T17:50:53.2586028Z .............................................i...................................................... 6700/9257
---
2019-11-20T17:56:00.9160271Z  finished in 5.589
2019-11-20T17:56:00.9334182Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T17:56:01.1197149Z 
2019-11-20T17:56:01.1197450Z running 156 tests
2019-11-20T17:56:04.0229031Z iiii....iii......iii..iiii...i.............................i..i..................i....i...........ii 100/156
2019-11-20T17:56:05.8936726Z .i.i..iiii..............i.........iii.i.........ii......
2019-11-20T17:56:05.8937275Z 
2019-11-20T17:56:05.8937416Z  finished in 4.960
2019-11-20T17:56:05.9122436Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T17:56:06.0793374Z 
---
2019-11-20T17:56:08.0237660Z  finished in 2.111
2019-11-20T17:56:08.0446945Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T17:56:08.2266834Z 
2019-11-20T17:56:08.2266973Z running 9 tests
2019-11-20T17:56:08.2267764Z iiiiiiiii
2019-11-20T17:56:08.2268363Z 
2019-11-20T17:56:08.2268411Z  finished in 0.183
2019-11-20T17:56:08.2458278Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T17:56:08.4272746Z 
---
2019-11-20T17:56:27.6555083Z  finished in 19.410
2019-11-20T17:56:27.6772305Z Check compiletest suite=debuginfo mode=debuginfo-gdb+lldb (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T17:56:27.8712793Z 
2019-11-20T17:56:27.8713052Z running 123 tests
2019-11-20T17:56:53.2097562Z .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....ii..........iiii..........i...ii...i.......ii. 100/123
2019-11-20T17:56:58.0850937Z i.i.i......iii.i.....ii
2019-11-20T17:56:58.0851426Z 
2019-11-20T17:56:58.0858207Z  finished in 30.408
2019-11-20T17:56:58.0867714Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T17:56:58.0869067Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2019-11-20T18:07:43.5168579Z 
2019-11-20T18:07:43.5246128Z  finished in 123.111
2019-11-20T18:07:43.5266358Z Testing core stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-20T18:07:43.7181960Z    Compiling core v0.0.0 (/checkout/src/libcore)
2019-11-20T18:07:49.3133390Z error[E0658]: use of unstable library feature 'iter_take_while_map': recently added
2019-11-20T18:07:49.3134437Z     --> src/libcore/../libcore/tests/iter.rs:1482:26
2019-11-20T18:07:49.3134860Z      |
2019-11-20T18:07:49.3135346Z 1482 |     assert_eq!(c.clone().take_while_map(|_| None).size_hint(), (0, None));
2019-11-20T18:07:49.3136191Z      |
2019-11-20T18:07:49.3136636Z      = help: add `#![feature(iter_take_while_map)]` to the crate attributes to enable
2019-11-20T18:07:49.3136837Z 
2019-11-20T18:07:49.3136837Z 
2019-11-20T18:07:49.3340493Z error[E0658]: use of unstable library feature 'iter_take_while_map': recently added
2019-11-20T18:07:49.3341399Z     --> src/libcore/../libcore/tests/iter.rs:1497:27
2019-11-20T18:07:49.3342222Z      |
2019-11-20T18:07:49.3342696Z 1497 |     assert_eq!(vi.clone().take_while_map(|_| None).size_hint(), (0, Some(10)));
2019-11-20T18:07:49.3343518Z      |
2019-11-20T18:07:49.3343958Z      = help: add `#![feature(iter_take_while_map)]` to the crate attributes to enable
2019-11-20T18:07:49.3344153Z 
2019-11-20T18:07:54.9193592Z error: aborting due to 2 previous errors
---
2019-11-20T18:07:55.0457970Z   local time: Wed Nov 20 18:07:55 UTC 2019
2019-11-20T18:07:55.3401698Z   network time: Wed, 20 Nov 2019 18:07:55 GMT
2019-11-20T18:07:55.3402586Z == end clock drift check ==
2019-11-20T18:07:55.9264418Z 
2019-11-20T18:07:55.9367554Z ##[error]Bash exited with code '1'.
2019-11-20T18:07:55.9403009Z ##[section]Starting: Checkout
2019-11-20T18:07:55.9404901Z ==============================================================================
2019-11-20T18:07:55.9404946Z Task         : Get sources
2019-11-20T18:07:55.9404984Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

WaffleLapkin commented:
Oh, I've forgotten that test crates are independent, and so they need feature-gate. (I have to run `./x.py` before committing argh >_<)

Is it better to add `#![feature(take_while_map)]` to `src/libcore/tests/iter.rs` or to remove tests added in 16327afc9587acf0306624be55e74e0225362aba?

withoutboats commented:
If we want to add this I think `map_while` is both clearer and briefer than `take_while_map`.

AnthonyMikh commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1609,6 +1609,100 @@ impl<I: Iterator, P> Iterator for TakeWhile<I, P>
+#[derive(Clone)]
+pub struct TakeWhileMap<I, P> {
+    iter: I,
+    flag: bool,
```
Maybe, not just flag but `finished`?

WaffleLapkin replied:
This field was copy-pasted from `TakeWhile` 🤷‍♀️ But I agree that `finished` is more clear than `flag` (actually when I was first reading `TakeWhile` sources I took me a while to understand the meaning of the `flag`)

AnthonyMikh commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1609,6 +1609,100 @@ impl<I: Iterator, P> Iterator for TakeWhile<I, P>
+    }
+}
+
+#[unstable(feature = "iter_take_while_map", reason = "recently added", issue = "0")]
```
Stability attributes currently don't work with trait impls

WaffleLapkin replied:
Why then [there is] a `#[stable(feature = "rust1", since = "1.0.0")]` on `Iterator` impl for `TakeWhile`?

[there is]: https://github.com/rust-lang/rust/pull/66577/files/ebd6bb2d56d1e57edf42ff9a2ee5306e4312f836#diff-36a07733644b558bd8371f796386234fR1552

dtolnay replied:
It's unstability attributes that don't work. ;) There is no such thing as an unstable trait impl.

WaffleLapkin replied:
I don't understand :( You said that there is no unstable trait impl, but
1) without `#[unsatble]` this code doesn't compile: ![image](https://user-images.githubusercontent.com/38225716/70691392-0ae56a00-1cca-11ea-8cb5-885e46b07361.png)
2) there are other `#[unstable]` impls, e.g.: [`impl TrustedLen for Rev`](https://github.com/rust-lang/rust/blob/ebd6bb2d56d1e57edf42ff9a2ee5306e4312f836/src/libcore/iter/adapters/mod.rs#L115-L117)
3) there are `#[stable]` impls too, e.g.: [`impl Iterator for Rev`](https://github.com/rust-lang/rust/blob/ebd6bb2d56d1e57edf42ff9a2ee5306e4312f836/src/libcore/iter/adapters/mod.rs#L39-L40)

AnthonyMikh commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1609,6 +1609,100 @@ impl<I: Iterator, P> Iterator for TakeWhile<I, P>
+                    None
+                }
+            }
+        }
```
Can be simpler:

```rust
if self.flag {
    return None
}

let x = self.iter.next()?;
let ret = (self.predicate)(x);
self.flag = ret.is_none();
ret
```

WaffleLapkin replied:
I think it would be even better to use `if+else` instead of `if+return`:
```rust
if self.flag {
    None
} else {
    let x = self.iter.next()?;
    let ret = (self.predicate)(x);
    self.flag = ret.is_none();
    ret
}
```

AnthonyMikh commented on src/libcore/iter/traits/iterator.rs:
```diff
@@ -1,10 +1,16 @@
+// ignore-tidy-filelength
+// This file almost exclusively consists of the definition of `Iterator`. We
+// can't split that into multiple files.
```
Well, technically we can split definition of Iterator into multiple files and `include!` them into source...

WaffleLapkin replied:
But we won't do that, right?

AnthonyMikh replied:
Well, that's up to libs team to decide

dtolnay replied:
I wouldn't want to do that. That seems worse.

JohnCSimon commented:
Ping from triage: 
@sfackler - Can you please review this PR?
Thanks.

sfackler commented:
@JohnCSimon this is still pending FCP waiting on @Kimundi, @KodrAus, @SimonSapin, @dtolnay, and @withoutboats.

withoutboats commented:
@rfcbot concern name

My view:

I don't have strong feelings either way about adding this PR, if someone else on the libs team wants to drive it (seems like sfackler does) that's fine with me, but if we do add it this API I want it to be named  `map_while`.

dtolnay commented:
+1 for calling this map_while.

@rfcbot reviewed

bors commented:
:umbrella: The latest upstream changes (presumably #67485) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #67540) made this pull request unmergeable. Please resolve the merge conflicts.

timvermeulen commented:
Heh, I'm not sure why my PR didn't gain the same traction this one does, but I'm glad it seems to be going somewhere. I actually almost submitted it as `take_while_map` too, before changing my mind 🙂

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=16692&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16692/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16692/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-24T01:20:02.3601948Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-24T01:20:02.3775248Z ##[command]git config gc.auto 0
2019-12-24T01:20:02.3859149Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-24T01:20:02.3917138Z ##[command]git config --get-all http.proxy
2019-12-24T01:20:02.4069016Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66577/merge:refs/remotes/pull/66577/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #67596) made this pull request unmergeable. Please resolve the merge conflicts.

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=16799&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16799/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/16799/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-12-25T18:06:58.5036800Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-12-25T18:06:58.5258474Z ##[command]git config gc.auto 0
2019-12-25T18:06:58.5335884Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-12-25T18:06:58.5405204Z ##[command]git config --get-all http.proxy
2019-12-25T18:06:58.5557666Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66577/merge:refs/remotes/pull/66577/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Dylan-DPC commented:
@WaffleLapkin can you fix the error? thanks

Mark-Simulacrum commented:
The API here still needs to be renamed to `map_while` per the fcpbot concern, so moving to waiting-on-author.

WaffleLapkin commented:
> @WaffleLapkin can you fix the error? thanks

Are you about https://github.com/rust-lang/rust/pull/66577#issuecomment-568920932? I've fixed it in 40bd6ef56eb27301919b4074b3a7122a7dd4571a

> The API here still needs to be renamed to map_while per the fcpbot concern, so moving to waiting-on-author.

Renamed :ok_hand:

Mark-Simulacrum commented:
@withoutboats concern on the name has been addressed (renaming to map_while). I don't personally think we need to wait for FCP to conclude here as this is just an unstable addition.

I have verified the implementation looks good to me.

@bors r+

bors commented:
:pushpin: Commit 8da929bcf7ae154e504cd17560665fe72d8bda6b has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 8da929bcf7ae154e504cd17560665fe72d8bda6b -->
<!-- homu: {"type":"Approved","sha":"8da929bcf7ae154e504cd17560665fe72d8bda6b","approver":"Mark-Simulacrum"} -->

Mark-Simulacrum commented:
Ah, actually -- let's get the commits squashed into one. r=me with that done.

@bors r-

lzutao commented:
Please also fix the PR title.

AnthonyMikh commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1752,6 +1752,95 @@ where
+#[unstable(feature = "iter_map_while", reason = "recently added", issue = "none")]
+impl<I: fmt::Debug, P> fmt::Debug for MapWhile<I, P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MapWhile").field("iter", &self.iter).field("flag", &self.finished).finish()
```
Does it make sense to expose a flag in debug output? Either the way, the name should be corrected to `finished`.

WaffleLapkin replied:
I'm not sure if it makes sense, but it's exposed in `Map`.

AnthonyMikh commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1752,6 +1752,95 @@ where
+}
+
+#[unstable(feature = "iter_map_while", reason = "recently added", issue = "none")]
+impl<B, I: Iterator, P> Iterator for MapWhile<I, P>
```
`MapWhile` should implement `FusedIterator`

WaffleLapkin replied:
Hmmm, I'm not sure why, but `TakeWhile` doesn't implement `DoubleEndedIterator`, are you sure `MapWhile` should implement it?

timvermeulen replied:
Neither `MapWhile` nor `TakeWhile` can implement `DoubleEndedIterator`, because you can only find the first element that doesn't satisfy the condition from the start.

WaffleLapkin replied:
@timvermeulen is it really required that condition must be satisfied from start? Why these examples shouldn't work?
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity).rev();

assert_eq!(iter.next(), Some(2));
assert_eq!(iter.next(), Some(1));
assert_eq!(iter.next(), None);
```
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity);

assert_eq!(iter.next_back(), Some(2));
assert_eq!(iter.next(), Some(0));
assert_eq!(iter.next_back(), Some(1));
assert_eq!(iter.next(), None);
```

timvermeulen replied:
Yes, a `DoubleEndedIterator` is required to yield the exact same items back-to-front as it does front-to-back (but obviously in reverse order). This is needed to be able to guarantee that `rev()` only ever reverses the order of the items, rather than yielding items you wouldn't otherwise have gotten.

WaffleLapkin replied:
I can't see that requirement nor in `DoubleEndedIterator`'s docs nor in `rev`'s docs. `DoubleEndedIterator`'s docs just says
> An iterator able to yield elements from both ends.

And that's all...

Also, if we'll decide to remove `finished flag`, then this requirement will be satisfied:
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity).rev();

assert_eq!(iter.next(), Some(2));
assert_eq!(iter.next(), Some(1));
assert_eq!(iter.next(), None);
assert_eq!(iter.next(), Some(0));
assert_eq!(iter.next(), None);
```
```rust
let vec = vec![Some(0), None, Some(1), Some(2)];
let mut iter = vec.into_iter().map_while(std::convert::identity);

assert_eq!(iter.next(), Some(0));
assert_eq!(iter.next(), None);
assert_eq!(iter.next(), Some(1));
assert_eq!(iter.next(), Some(2));
assert_eq!(iter.next(), None);
```

timvermeulen replied:
> I can't see that requirement nor in DoubleEndedIterator's docs nor in rev's docs.

I agree that the docs could be a little clearer about this. Nonetheless, I think it goes without saying that it would be highly unexpected if for some generic iterator type `I` calling `rev()` on it would give you completely different items. It wouldn't be very useful, either – you could already call `rev()` before calling `map_while` to achieve the same thing.

I'm not sure what you mean with the first snippet, `MapWhile` doesn't implement `DoubleEndedIterator` so we can't call `rev()` on it in the first place. As for the second snippet, it is perfectly fine behavior to return `Some` after having returned `None` as long as `MapWhile` doesn't implement `FusedIterator`. That's the whole reason `fuse` exists. All that the `finished` flag does is make the iterator fused, which is not the responsibility of an individual iterator.

WaffleLapkin replied:
> I'm not sure what you mean with the first snippet, MapWhile doesn't implement DoubleEndedIterator so we can't call rev() on it in the first place. 

With both snippets, I mean that we can (and should) implement `DoubleEndedIterator` for `MapWhile`. It wouldn't "give you completely different items", it would give you the same items in reversed order, as expected.

By the way, [`Take`](https://doc.rust-lang.org/std/iter/struct.Take.html) implements `DoubleEndedIterator` and it could return different results if limit (`n`) < length of the inner iterator.

timvermeulen replied:
Ah sorry, I misunderstood the point you were making. Things like a `for` loop and `collect` will only call `next` until a `None` is encountered, so those two iterators don't actually produce the same values. The first one produces `2` and `1`, and the second one only produces `0`. What follows after `None` is irrelevant. Therefore, the last element of the (non-reversed) iterator is `1`, and you can only know that by iterating from the start.

`Take` doesn't have this problem because given `n` and the length of the iterator, you can know how many elements to skip from the end to get to the last element (note that `Take` only implements `DoubleEndedIterator` when the inner iterator is an `ExactSizeIterator`, and `Take::next_back` calls `self.iter.nth_back(...)`).

WaffleLapkin replied:
Oh, I didn't notice how `DoubleEndedIterator` is implemented for `Take`. Thanks for explanations!

Mark-Simulacrum commented:
Let's move further improvements to a future PR (this is unstable, so doing so is definitely fine), to avoid giving pings to everyone and so forth.

@bors r+

bors commented:
:pushpin: Commit 47e07b30f80c64dcc6729c3840833ea7b4515606 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 47e07b30f80c64dcc6729c3840833ea7b4515606 -->
<!-- homu: {"type":"Approved","sha":"47e07b30f80c64dcc6729c3840833ea7b4515606","approver":"Mark-Simulacrum"} -->

Mark-Simulacrum commented:
Oh, wait, no, this still needs a tracking issue. I've filed https://github.com/rust-lang/rust/issues/68537, please use that in the `#[unstable(...)]` annotations.

@bors r-

WaffleLapkin commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1752,6 +1752,95 @@ where
+#[derive(Clone)]
+pub struct MapWhile<I, P> {
+    iter: I,
+    finished: bool,
```
The whole point of the `MapWhile` is that it stops on first `None` just like `TakeWhile` stops on first `false`

ollie27 replied:
It stops by having `.next()` return `None`. The `finished` field isn't needed to do that. After then it doesn't matter what `.next()` returns unless the iterator implements `FusedIterator` which `MapWhile` doesn't.

timvermeulen replied:
I first thought it was to improve the upper bound of `size_hint`, but it also doesn't matter what that method returns once the iterator has been exhausted. So I suppose you're right. We probably can't remove it from `TakeWhile` though, since it already implements `FusedIterator`.

WaffleLapkin replied:
It can be a bit confusing, so if we do remove `finished` flag, then I think we need to mention it in the docs

WaffleLapkin commented on src/libcore/iter/adapters/mod.rs:
```diff
@@ -1752,6 +1752,95 @@ where
+}
+
 #[stable(feature = "fused", since = "1.26.0")]
 impl<I, P> FusedIterator for TakeWhile<I, P>
```
Yeah, I've also noticed it recently. I'll fix it in following pr(s?)

bors commented:
:umbrella: The latest upstream changes (presumably #68234) made this pull request unmergeable. Please resolve the merge conflicts.

WaffleLapkin commented:
Rebased to resolve conflicts with master

WaffleLapkin commented:
@Mark-Simulacrum  Can you merge this pr? I've added issue to the `#[unstable(...)]` annotations

Dylan-DPC commented:
@bors r=mark-simulcrum

bors commented:
:pushpin: Commit db1a107b3f920637dc785fcc6d6bbe247a271e7b has been approved by `mark-simulcrum`

<!-- @bors r=mark-simulcrum db1a107b3f920637dc785fcc6d6bbe247a271e7b -->
<!-- homu: {"type":"Approved","sha":"db1a107b3f920637dc785fcc6d6bbe247a271e7b","approver":"mark-simulcrum"} -->

bors commented:
:hourglass: Testing commit db1a107b3f920637dc785fcc6d6bbe247a271e7b with merge 34700c15c7a49d400951da385b2eacab6fb6fe86...
<!-- homu: {"type":"BuildStarted","head_sha":"db1a107b3f920637dc785fcc6d6bbe247a271e7b","merge_sha":"34700c15c7a49d400951da385b2eacab6fb6fe86"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=19756)
Approved by: mark-simulcrum
Pushing 34700c15c7a49d400951da385b2eacab6fb6fe86 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"mark-simulcrum","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=19756"},"merge_sha":"34700c15c7a49d400951da385b2eacab6fb6fe86"} -->
